File: daily/1497.js
const fs = require("fs");
const inputPath = process.platform === "linux" ? "/dev/stdin" : "./input.txt";
const stdin = fs
  .readFileSync(inputPath)
  .toString()
  .split("\n")
  .map((line) => line.trim());

const input = (() => {
  let line = 0;
  return () => stdin[line++];
})();

const solution = () => {
  const [n, m] = input().split(" ").map(Number);

  const g = Array.from({ length: n }, () => {
    const [name, songs] = input().split(" ");
    let bin = 0;
    for (let i = 0; i < m; i++) {
      if (songs[i] === "Y") bin |= 1 << i;
    }
    return [name, bin];
  });
  let maxSong = 0;
  let ans = -1;
  const countG = (num) => {
    let cnt = 0;
    while (num) {
      cnt += num & 1;
      num >>= 1;
    }
    return cnt;
  };

  const recur = (cnt, state, index) => {
    const nowSongCount = countG(state);

    if (nowSongCount > maxSong) {
      maxSong = nowSongCount;
      ans = cnt;
    } else if (nowSongCount === maxSong) {
      if (ans > cnt) {
        ans = cnt;
      }
    }

    if (index === n) return;

    recur(cnt + 1, state | g[index][1], index + 1);
    recur(cnt, state, index + 1);
  };

  recur(0, 0, 0);
  console.log(ans);
};

solution();

File: daily/1497.py
from re import L
import sys
input = sys.stdin.readline

n, m = map(int, input().split())
info = []
max_ = 0
ans = -1

for _ in range(n):
  g, s  = input().split()
  temp = 0
  for i in range(len(s)):
    if (s[i] == "Y"):
      temp |= 1 << i
  info.append((g, temp))

def bit_counting(bit):
  ret = 0;
  while bit:
    ret += bit & 1
    bit >>= 1
  return ret

def recur(cnt, depth, state):
  global max_, ans
  now = bit_counting(state)
  if now > max_:
    max_ = now
    ans = cnt
  elif now == max_ :
    ans = min(ans, cnt)

  if depth == n: return

  recur(cnt + 1, depth + 1, state | info[depth][1])
  recur(cnt, depth + 1, state)

recur(0, 0, 0)
print(ans)